Generate mailer files in auth generator only if ActionMailer is used

ramimassoud opened the pull request:
### Motivation / Background

This Pull Request has been created because if the authentication generator is run in an app created with `--skip-action-mailer`, the app will be silently broken. When deployed to an environment where `eager_load` is true, the app will fail to start as the generated `PasswordMailer` inherits from a nonexistent `ApplicationMailer`.

Fixes #54501 

### Detail

This Pull Request changes the authentication generator to skip all of the mailer-related files if `ActionMailer` is not defined. Also leave `PasswordsController#create` as blank to be implemented by the user.

### Additional information

In my use case, a simple login was enough, and I created user(s) manually in a console. I do not need a password reset flow. An alternative approach would be to require an app to have `ActionMailer` to run the generator.

### Checklist

Before submitting the PR make sure the following are checked:

* [x] This Pull Request is related to one change. Unrelated changes should be opened in separate PRs.
* [x] Commit message has a detailed description of what changed and why. If this PR fixes a related issue include it in the commit message. Ex: `[Fix #issue-number]`
* [x] Tests are added or updated if you fix a bug or add a feature.
* [x] CHANGELOG files are updated for the changed libraries if there is a behavior change or additional feature. Minor bug fixes and documentation changes should not be included.

byroot commented on railties/lib/rails/generators/rails/authentication/templates/app/controllers/passwords_controller.rb.tt:
```diff
@@ -7,11 +7,13 @@ class PasswordsController < ApplicationController
   end
 
   def create
+  <% if defined?(ActionMailer::Railtie) -%>
```
Shouldn't this action just not exist if ActionMailer is skipped? It's basically a noop.

The route would probably need to be removed to?

ramimassoud replied:
Ah yes, of course! I'll take a crack at adjusting the routes and removing `#create` and `#new`.

One thing I noticed is the way the routes are set up would also generate `#show`, `#index`, and `#destory`. Would it make sense to always have an `only:` in there? Then if the mailer is defined it would be `only: [:new, :create, :edit, :update]`, otherwise `only: [:edit, :update]`

https://github.com/rails/rails/blob/7cb69a8fd004c94b9266f2fdb2a833d15c09192b/railties/lib/rails/generators/rails/authentication/authentication_generator.rb#L40-L43

skipkayhil replied:
I think adding `only:` to generated routes would not be accepted: https://github.com/rails/rails/pull/52369#discussion_r1699079086

ramimassoud replied:
@skipkayhil Thank you for the heads up! 🙏 

@byroot I think I addressed the feedback fully. Let me know if you think any more changes would be appropriate here.

ramimassoud committed:
Generate mailer files in auth generator only if ActionMailer is used

If an app does not use ActionMailer, then all of the mailer-related
code is unnecessary. Also if the mailer files are generated, the
PasswordMailer will inherit from an ApplicationMailer that doesn't
exist, breaking the application in a way that is not obvious until the
mailer is run or the app is deployed to an environment with eager
class loading

Fixes #54501
